contracts/CarryPublicTokenCrowdsale.sol

```diff
@@ -0,0 +1,149 @@
+    IndividualMaxCap[] public individualMaxCaps;
+
+    mapping(address => uint256) public contributions;
+
+    uint256 public closingTime;
+
+    constructor(
+        address _wallet,
+        CarryToken _token,
+        uint256 _rate,
+        uint256 _cap,
+        uint256 _closingTime,
+        uint256 _individualMinPurchaseWei,
+
+        // Since Solidity currently doesn't allows parameters to take array of
+        // structs, we work around this by taking (timestamp, weis) pairs as
+        // a 1d-array of [timestamp1, weis1, timestamp2, weis2, ...].
+        // It fails if the length is not even but odd.
+        uint256[] _individualMaxCaps
+    ) public CappedCrowdsale(_cap) Crowdsale(_rate, _wallet, _token) {
+        require(
+            _individualMaxCaps.length % 2 == 0,
+            "The length of _individualMaxCaps has to be even, not odd."
+        );
+        closingTime = _closingTime;
+        individualMinPurchaseWei = _individualMinPurchaseWei;
+        for (uint256 i = 0; i < _individualMaxCaps.length; i += 2) {
+            individualMaxCaps.push(
+                IndividualMaxCap(
+                    _individualMaxCaps[i],
+                    _individualMaxCaps[i + 1]
+                )
+            );
+        }
+    }
+
+    function _preValidatePurchase(
+        address _beneficiary,
+        uint256 _weiAmount
+    ) internal whenNotPaused {
+        // Prevent gas war among purchasers.
+        require(
+            tx.gasprice <= maxGasPrice,
+            "Gas price is too expensive. Don't be competitive."
+        );
+
+        super._preValidatePurchase(_beneficiary, _weiAmount);
+
+        // Tokensale is closed after closingTime.
+        require(
+            // solium-disable-next-line security/no-block-members
+            block.timestamp <= closingTime,
+            "Already closed; it's too late."
+        );
+
+        uint256 contribution = contributions[_beneficiary];
+        uint256 contributionAfterPurchase = contribution.add(_weiAmount);
+
+        // If a contributor already has purchased a minimum amount, say 0.1 ETH,
+        // then they can purchase once again with less than a minimum amount,
+        // say 0.01 ETH, because they have already satisfied the minimum
+        // purchase.
+        require(
+            contributionAfterPurchase >= individualMinPurchaseWei,
+            "Sent ethers is not enough."
+        );
+
+        // See also the comment on the individualMaxCaps above.
+        uint256 latestTimestamp = 0;
+        uint256 individualMaxWei = 0;
+        for (uint256 i = 0; i < individualMaxCaps.length; i++) {
+            // solium-disable-next-line security/no-block-members
+            if (individualMaxCaps[i].timestamp <= block.timestamp &&
+                individualMaxCaps[i].timestamp > latestTimestamp) {
+                latestTimestamp = individualMaxCaps[i].timestamp;
+                individualMaxWei = individualMaxCaps[i].maxWei;
```

Comment:
Can we assume latest `maxWei` is `individualMaxWei`? 

e.g.
```
individualMaxCaps = [
    IndividualMaxCap(1533081600000, 10 ether),
    IndividualMaxCap(1533686400000, 5 ehter)
]
```

Reply:
@longfin It was actually intended.  I thought the individual max cap should be possible to be overridden by the cap of a less amount (as well as a more amount).  By the way, indeed this behavior could supplant `closingTime` as well.  How do you see this?

Reply:
looks good